src/wildcard.cpp

```diff
@@ -435,7 +435,16 @@ static const wchar_t *file_get_desc(int lstat_res, const struct stat &lbuf, int
 /// up. Note that the filename came from a readdir() call, so we know it exists.
 static bool wildcard_test_flags_then_complete(const wcstring &filepath, const wcstring &filename,
                                               const wchar_t *wc, expand_flags_t expand_flags,
-                                              completion_receiver_t *out) {
+                                              completion_receiver_t *out, bool known_dir) {
+    const bool executables_only = expand_flags & expand_flag::executables_only;
+    const bool need_directory = expand_flags & expand_flag::directories_only;
+    // Fast path: If we need directories, and we already know it is one,
```

Comment:
You can add another fast path: if we are `executables_only` and `known_dir` immediately return false!